salt/_pillar/metalk8s_nodes.py

```diff
@@ -50,13 +52,92 @@ def node_info(node, ca_minion):
     return result
 
 
+def get_cluster_version(api_client):
+
+    coreV1 = kubernetes.client.CoreV1Api(api_client=api_client)
+    namespace = coreV1.read_namespace(name="kube-system")
+    annotation_key = 'metalk8s.scality.com/cluster-version'
+
+    if annotation_key not in namespace.metadata.annotations:
+        return __utils__['pillar_utils.errors_to_dict'](
+            'Unable to retrieve cluster version'
+        )
+
+    return namespace.metadata.annotations[annotation_key]
+
+
+def iso_timestamp_converter(timestamp):
+    if not timestamp:
+        return None
+    return timestamp.isoformat()
+
+
+def get_storage_classes(api_client):
+    storageV1 = kubernetes.client.StorageV1Api(api_client=api_client)
+    storage_classes = {}
+    for storageclass in storageV1.list_storage_class().items:
+        # Need to convert the datetime object in storageclass to ISO format in
+        # order to make them serializable.
+        storageclass.metadata.creation_timestamp = iso_timestamp_converter(
+            storageclass.metadata.creation_timestamp
+        )
+        storageclass.metadata.deletion_timestamp = iso_timestamp_converter(
+            storageclass.metadata.deletion_timestamp
+        )
+        storage_classes[storageclass.metadata.name] = storageclass.to_dict()
+
+    return storage_classes
+
+
+def list_volumes(api_client, minion_id):
+    results = {}
+    customObjectApi = kubernetes.client.CustomObjectsApi(api_client=api_client)
+    storage_classes = get_storage_classes(api_client)
+
+    if not storage_classes:
+        return __utils__['pillar_utils.errors_to_dict'](
+            'Cannot list storage classes'
+        )
+
+    try:
+        volumes = customObjectApi.list_cluster_custom_object(
+            group="storage.metalk8s.scality.com",
+            version="v1alpha1",
+            plural="volumes"
+        )
+    except ApiException as exc:
+        return __utils__['pillar_utils.errors_to_dict'](
+            'Cannot get volumes {}'.format(exc)
+        )
+
+    for volume in volumes['items']:
+        name = volume['metadata']['name']
+
+        if volume['spec']['nodeName'] != minion_id:
+            continue
+        storageclass = storage_classes.get(
+            volume['spec'].pop('storageClassName')
+        )
+        if not storageclass:
+            results[volume['metadata']['name']] = \
+                'invalid StorageClass. List of storage classes {}'.format(
+                    list(storage_classes.keys())
+                )
+            continue
+        volume['spec']['storageClass'] = storageclass
+
+        results[name] = volume
+    return results
+
+
 def ext_pillar(minion_id, pillar, kubeconfig):
     if not os.path.isfile(kubeconfig):
```

Comment:
I'd rather have:

```python
if not os.path.isfile(kubeconfig):
    error_tplt = '...'
    error_dict = __utils__['pillar_utils.errors_to_dict'](...)
    return {'metalk8s': error_dict}
```

and continue from there, instead of assigning this error dict to some variables we re-use later.

Reply:
I am not sure of the consequences this will have, but looking through the repo on how we generally handle errors with external pillar, I see similar patterns.

I am willing to discuss this more with salt experts herein.

This issue is being tracked by #1498